ICML writer: add FirstParagraph and Bibliography styles.

jgm opened the pull request:
Closes #11268.

Not sure this is the right approach. Would it be better to add FirstParagraph *in addition to* Paragraph?  (And similarly for Bibliography?)  If the styles are nestable in this way (I don't know a thing about ICML), then this would be less disruptive, as people who have customized the Paragraph style would not need to do anything special when using the new writer.

Alternatively, perhaps we could have the style define the FirstParagraph and Bibliography styles in terms of Paragraph.  (But this is less ideal for various reasons.)

jgm committed:
ICML writer: add FirstParagraph and Bibliography styles.

Closes #11268.

hevonen commented:
Hmm I see (and maybe finally understand too), but isn't that just a single style which is named to look like two styles from XML point of view? There is a specific style with matching name in ICML root styles section:
```
<ParagraphStyle Self="ParagraphStyle/Blockquote &gt; Paragraph" Name="Blockquote &gt; Paragraph" LeftIndent="10">
  <Properties>
    <BasedOn type="object">$ID/NormalParagraphStyle</BasedOn>
  </Properties>
</ParagraphStyle>
```

Pandoc's multiple styles are a single style in ICML (a key-value pair). Every used style needs to also exist in ICML's root styles section to work. If e.g. `"ParagraphStyle/Paragraph &gt; first"` does not exist there it is undefined and paragraphs referring to it will default to [Basic Paragraph] in Indesign. Indesign does not do style combinations from a list of styles. It reads the "list of styles" as a specific style name. Because of this these styles look slightly weird: "Blockquote > Paragraph" instead of "Blockquote" (styles can be renamed but if linked ICML is updated the old styles appear again and renamed styles go unused).

So while "Name" and "Self" attribute values can be a list of styles that list is a single string identifier in ICML. Working example that mimics this is attached.

It would be nice if the new first paragraph root style would have BasedOn tag with "ParagraphStyle/Paragraph" to keep existing layouts the same (as in attached example ICML). But for me getting first paragraphs and bibliography tagged is more important as that change is quick to do in Indesign.

[pandoctest-1st-paragraph-and-bibliography-v3.icml.txt](https://github.com/user-attachments/files/23392933/pandoctest-1st-paragraph-and-bibliography-v3.icml.txt)

jgm commented:
Hm. Wouldn't we also need to say what`Blockquote > FirstParagraph` is based on? (And so on?)

jgm committed:
ICML writer: base FirstParagraph styles on corresponding...

Paragraph styles.  This should ensure that existing documents
still work with the FirstParagraph style, without the need
to define new styles.

hevonen commented:
Thanks, I managed to compile it and it works! I tried with ≈160k character document and at least no contents were lost compared to ordinary pandoc.

I'm wondering what are the best practises for style naming and generation.

Current version emits FirstParagraph to every style including custom styles. This can lead to somewhat messy set of styles compared to e.g. docx export. Docx styles are also named differently (e.g. "Body Text" and "First Paragraph" vs "Paragraph" and "FirstParagraph") but I guess that pandoc does not have codified style names for different writers?

Currently bullet list styles have their first item style named as "BulList > first". Would it be a good practise to name FirstParagraph similarly "Paragraph > first"? And maybe omit "FirstParagraph" and "Paragraph" for bibliography and possibly for custom styles (if it is needed it could be defined manually)? Bibliography could just be "Bibliography"? Also, would it be feasible to generate "FirstParagraph" after forced empty line (to me empty space indicates a new section of text even without heading)?

Also, current FirstParagraphs don't have base style set (maybe not a problem).

Would it be possible to skip some of these styles (or alternatively enable them) e.g. in metadata block? That might be a bad fit for other document sources than Markdown and adding command line options for this is probably not wanted either.

Example of pandoc styles for markdown with some custom styles ("Basic Paragraph" is Indesign's base style):
```
[Basic Paragraph]
BulList
BulList > first
Footnote > Paragraph
Header1
Headerl (unnumbered)
Header2
Header3
Paragraph
author > Paragraph
quote > Blockquote > Paragraph
quote > Paragraph
quoteauthor > Blockquote > Paragraph
quoteauthor > Paragraph
thanks > BulList
thanks > BulList > first
thanks > Header2
thanks > Paragraph
title > Header1
Bibliography > Paragraph
Bibliography > FirstParagraph
FirstParagraph
author > FirstParagraph
quote > FirstParagraph
quoteauthor > Blockquote > FirstParagraph
thanks > FirstParagraph

```

sjardim commented:
> We do use multiple styles, though. E.g. for block quotes.
> 
> ```
> <ParagraphStyleRange AppliedParagraphStyle="ParagraphStyle/Blockquote &gt; Paragraph">
>   <CharacterStyleRange AppliedCharacterStyle="$ID/NormalCharacterStyle">
>     <Content>hi</Content>
>   </CharacterStyleRange>
> </ParagraphStyleRange>
> ```

Graphic designer here chimeing in to clarify: we can have multiple CHARACTER styles applied to parts of the string, but ONLY ONE Paragraph style applied to a paragraph. So this quoted example works fine inside InDesign. 

I'd love a way to create a style_mapping file and pass it to Pandoc so it could match its internal style to mine, example:

`Bullist > Paragraph` would match with "Bulleted List" in my style_mapping file and so on. This is the dream, as I could get an Indesign file from another designer with their styles already there and just copy them to my mapping. :)
